pkg/daemon/ceph/osd/kms/kms.go

```diff
@@ -121,7 +135,14 @@ func (c *Config) GetSecret(secretName string) (string, error) {
 func (c *Config) DeleteSecret(secretName string) error {
 	if c.IsVault() {
 		// Store the secret in Vault
-		v, err := InitVault(c.context, c.clusterInfo.Namespace, c.clusterSpec.Security.KeyManagementService.ConnectionDetails)
+		removeCertFiles := make(chan bool, 1)
+		// Remove cert files from operator's filesystem
+		defer func() {
+			removeCertFiles <- true
+			close(removeCertFiles)
+		}()
+
+		v, err := InitVault(c.context, c.clusterInfo.Namespace, c.clusterSpec.Security.KeyManagementService.ConnectionDetails, removeCertFiles)
 		if err != nil {
 			return errors.Wrap(err, "failed to delete secret in vault")
 		}
```

Comment:
I want to make sure my understanding is correct.

1. The defer means that the files need removed once this function call is complete (this also goes for the other functions that use the defer like this.
2. The func `InitVault` starts a goroutine that waits on the `removeCertFiles` channel to show `true` to then delete the files.

---- 

Some questions about this implementation:
1. Why does the goroutine in `InitVault` need to watch for `true` when it could just wait for the channel to be closed? I wonder if looking for the value when the channel will immediately be closed after will cause any race conditions.
2. If the files are always removed after the function is done running, is it a cleaner implementation to not start another goroutine and set up a channel and instead return a `RemoveCertFiles` function from `InitVault` that we can defer for afterwards? That way, we can verify if the files are removed correctly, and we don't have to mess with the complication of channels? (example below)

```go
type RemoveCertFilesFunction func() error

func InitVault(/* vars */) (RemoveCertFilesFunction, error) {
    /* do stuff */
    return RemoveCertFilesFunction() {
        for _, file := range files {
            /* delete file */
        }, nil
}

func (c *Config) DeleteSecret(secretName string) error {
    removeCertFiles, err := InitVault(/* vars */)
    if err != nil {
        return err
    }
    defer removeCertFiles()
    
    /* other stuff */
    
    return nil
}

```

Reply:
Agreed, I am wondering why the files can't just always be removed immediately. What is the need for the approach with goroutines?

Reply:
Ok so I think I went too deep into the weeds. The certificates are written as files on the operator's filesystem so that they can be loaded by `tls.LoadX509KeyPair()`, once finished we can get rid of the files. Thanks, @BlaineEXE for the alternate approach, much simpler at the end of the day, with a slight change in implementation.